src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/Target_X64/X64ReadyToRunHelperNode.cs

```diff
@@ -225,5 +225,98 @@ protected override void EmitCode(NodeFactory factory, ref X64Emitter encoder, bo
                     throw new NotImplementedException();
             }
         }
+
+        // emits code that results in ThreadStaticBase referenced in RAX.
+        // may trash volatile registers. (there are calls to the slow helper and possibly to platform's TLS support)
+        private static void EmitInlineTLSAccess(NodeFactory factory, ref X64Emitter encoder)
+        {
+            ISymbolNode getInlinedThreadStaticBaseSlow = factory.HelperEntrypoint(HelperEntrypoint.GetInlinedThreadStaticBaseSlow);
+            ISymbolNode tlsRoot = factory.TlsRoot;
+            bool singleFileExe = factory.CompilationModuleGroup.IsSingleFileCompilation;
```

Comment:
We can emit slightly code when we know it is going to be a single exe. This is particularly useful on Windows, since there are not linker relaxations, but is mildly advantageous on Linux as well.

Reply:
Is this assuming that `_tls_index` is going to be zero when `IsSingleFileCompilation` is set? I do not think it is safe assumption to make. `IsSingleFileCompilation` does not guarantee that the output is .exe.

I am not sure whether this is a safe assumption to make even for .exes - is that documented somewhere or have you seen any Windows C++ compiler doing this optimization?

Reply:
_tls_index would be 0 in the most common case that we support right now where the TLS variable is used from the same module that defines it and it is known at link time. I am not sure if `IsSingleFileCompilation` guarantees that, but it would be useful if we could know if we have such case as inlining shorter code is attractive.  
Overall - Since in such cases _tls_index will be 0 constant at link time, I think this is a minor optimization since it mostly just makes the code shorter. It does not reduce the number of memory accesses or indirection. It is not a big deal if we can't do this reliably.

C++ compiler does this optimization when[ /GA switch is used](https://learn.microsoft.com/en-us/cpp/build/reference/ga-optimize-for-windows-application?view=msvc-170) (optimize for application), but it is also known that it may cause trouble in rare cases:

> It’s worth noting that in the (extremely rare) case that a .exe exports functions and is imported by another .exe, this optimization will cause random corruption if the imported .exe happens to use __declspec(thread).

Source: http://www.nynaeve.net/?p=185

Reply:
I did not know about `/GA` switch. `IsSingleFileCompilation` does not have the right semantics for this. We would need a new compiler switch that says we are building an application.

Reply:
The same can be said about Linux case. GD form is the safest to emit and linker will relax it to LS form if that is ok.

Emitting an `Initial Exec` model is mostly size saving, but it could be attractive, considering it is a very common case for us and having smaller code is nice when inlining.  
The applicability of the model is defined slightly different than on Windows, but I think it means roughly the same - not dynamically loaded and not cross-module accessed.

> A more restrictive optimization is usable if the variables accessed are known to be in one of the modules available at program start and if the programmer selects to use the static access model. 

It is a good option if we can do this, but emitting GD form is not that much worse to be too worried about.

Reply:
Yes, this optimization can be enabled on both Windows and Unix, via a new compiler switch.

Reply:
To put things in the perspective - for Linux the `Initial Exec` sequence is the same 16 bytes, but does not need an 
additional `mov rdi, rax` after that as there is more flexibility where the result would be if you do not imply a possible method call.  

It saves some, but not a huge saving overall.

Reply:
I have short-circuited the "Initial Exec" optimizations, but left the code, assuming we will implement a compiler switch similar to `/GA`.

Reply:
TLS Relocations are more complex on linux-arm64 for the dynamic case while LE case is still quite simple. It makes the support for switch like  `\GA` / `-fno-pic -fpie` a lot more interesting.